src/wallet/wallet2.cpp

```diff
@@ -1602,6 +1602,46 @@ std::string wallet2::get_subaddress_label(const cryptonote::subaddress_index& in
   return m_subaddress_labels[index.major][index.minor];
 }
 //----------------------------------------------------------------------------------------------------
+void wallet2::scan_tx(std::vector<std::string> txids)
+{
+  // If keys are encrypted, ask for password now to avoid timing leak when finding first output.
+  if (m_ask_password == AskPasswordToDecrypt && !m_unattended && !m_watch_only )//&& !m_multisig_rescan_k)
+  {
+    static critical_section password_lock;
+    CRITICAL_REGION_LOCAL(password_lock);
+    if (!m_encrypt_keys_after_refresh)
+    {
+        boost::optional<epee::wipeable_string> pwd = m_callback->on_get_password("scan_tx");
+        THROW_WALLET_EXCEPTION_IF(!pwd, error::password_needed, tr("Password is needed to scan transactions"));
+        THROW_WALLET_EXCEPTION_IF(!verify_password(*pwd), error::password_needed, tr("Invalid password: password is needed to scan transactions"));
+        decrypt_keys(*pwd);
+        m_encrypt_keys_after_refresh = *pwd;
+    }
+  }
+
+  for (std::string txid : txids) {
+    cryptonote::COMMAND_RPC_GET_TRANSACTIONS::request req = AUTO_VAL_INIT(req);
+    cryptonote::COMMAND_RPC_GET_TRANSACTIONS::response res = AUTO_VAL_INIT(res);
+    req.decode_as_json = false;
+    req.prune = true;
+    req.txs_hashes.push_back(txid);
+
+    {
+      const boost::lock_guard<boost::recursive_mutex> lock{m_daemon_rpc_mutex};
+      req.client = get_client_signature();
+      bool r = epee::net_utils::invoke_http_json("/gettransactions", req, res, *m_http_client, rpc_timeout);
+      THROW_WALLET_EXCEPTION_IF(!r, error::wallet_internal_error, "Failed to get transaction from daemon");
+      THROW_WALLET_EXCEPTION_IF(res.txs.size() != 1, error::wallet_internal_error, "Failed to get transaction from daemon");
+    }
+
+    cryptonote::transaction tx;
+    crypto::hash tx_hash;
+    THROW_WALLET_EXCEPTION_IF(!get_pruned_tx(res.txs.front(), tx, tx_hash), error::wallet_internal_error, "Failed to get transaction from daemon");
+
+    process_new_transaction(tx_hash, tx, res.txs.front().output_indices, res.txs.front().block_height, /*block_version=*/0, res.txs.front().block_timestamp, /*miner_tx=*/false, res.txs.front().in_pool, res.txs.front().double_spend_seen, {}, {});
+  }
```

Comment:
Why not request all transactions at once?

Reply:
I was unsure about the max transactions allowed per request and decided to do this as first approach. Should I change it and request them all at once?

Reply:
And if it fails... batch them in smaller batches?

Reply:
Request `RESTRICTED_TRANSACTIONS_COUNT` transactions per request. Split into multiple requests if needed.

Reply:
I'll post a commit with this tomorrow. I have to go now. Thanks for the help.